includes/core/server/class-request.php

```diff
@@ -0,0 +1,531 @@
+	 * @param WC_Payments_Http_Interface $http_interface The HTTP interface for the server.
+	 */
+	public function __construct( WC_Payments_API_Client $api_client, WC_Payments_Http_Interface $http_interface ) {
+		$this->api_client     = $api_client;
+		$this->http_interface = $http_interface;
+	}
+
+	/**
+	 * Returns the needed API.
+	 *
+	 * @return string Check WCPay\Core\Server\APIs.
+	 */
+	abstract public function get_api(): string;
+
+	/**
+	 * Returns the method of the request.
+	 *
+	 * @return string See the constants in WordPress's `Requests` class.
+	 */
+	abstract public function get_method(): string;
+
+	/**
+	 * This is a legacy method, and is the same throughout the codebase.
+	 * Might be worth removing while refactoring to use the Core\Server API.
+	 *
+	 * @return bool
+	 */
+	public function is_site_specific(): bool {
+		return true;
+	}
+
+	/**
+	 * If true, the request will be signed with the user token rather than blog token. Defaults to false.
+	 *
+	 * @return bool
+	 */
+	public function should_use_user_token(): bool {
+		return false;
+	}
+
+	/**
+	 * Indicates if the raw response should be returned.
+	 *
+	 * @return bool
+	 */
+	public function should_return_raw_response(): bool {
+		return false;
+	}
+
+	/**
+	 * Returns all of the parameters for the request.
+	 *
+	 * @return array
+	 * @throws Invalid_Request_Parameter_Exception If the request has not been initialized yet.
+	 */
+	final public function get_params() {
+		$defaults = $this->get_default_params();
+		$params   = array_merge( $defaults, $this->params );
+
+		$missing_params = [];
+		foreach ( $this->get_required_params() as $name ) {
+			if ( ! isset( $params[ $name ] ) ) {
+				$missing_params[] = $name;
+			}
+		}
+
+		if ( ! empty( $missing_params ) ) {
+			throw new Invalid_Request_Parameter_Exception(
+				sprintf(
+					'Trying to access the parameters of a request which is not (fully) initialized yet. Missing parameter(s) for %s: %s',
+					get_class( $this ),
+					implode( ', ', $missing_params )
+				),
+				'wcpay_core_invalid_request_parameter_missing_parameters'
+			);
+		}
+
+		foreach ( $params as $key => $value ) {
+			if ( true === $value ) {
+				// The WCPay server requires the string 'true'.
+				$params[ $key ] = 'true';
+			}
+		}
+
+		return $params;
+	}
+
+	/**
+	 * Get request param by key.
+	 *
+	 * @param string $key Key to get.
+	 *
+	 * @return mixed
+	 * @throws Invalid_Request_Parameter_Exception
+	 */
+	final public function get_param( $key ) {
+		if ( array_key_exists( $key, $this->params ) ) {
+			return $this->params[ $key ];
+		}
+		throw new Invalid_Request_Parameter_Exception( 'The passed key does not exist in Request class', 'wcpay_core_invalid_request_parameter_uninitialized_param' );
+	}
+
+	/**
+	 * Allows the request to be executed and formatted.
+	 *
+	 * @return mixed          Either the response array, or the correct object.
+	 */
+	final public function send() {
```

Comment:
> sometimes you won't need any filters to pass, and you will not be able to execute the request.

Wait, you do realize filters (and hooks in general, incl actions) are optional, right? By providing the hook, you allow somebody (incl. 3rd-party devs) to use it if/whenever it is needed. The more hooks a plugin has, the easier it is to adjust its behavior to suit your needs without having to change its code. That's why I am trying to enforce every request to go through a filter like this, and making it optional, renders the point moot.

Reply:
I am aware that filters are optional, but the way you describe their usages is a way where you enforce it. Adding a filter for a sake of adding it doesn't help too much either. What if you need only request without any filters? With the approach, you mentioned here and with the required parameter in `send()` function make things a bit complicated. With this approach, you are using getters and setters to build the request and after that, you create a filter that will do something and in the cases where you need plain request class, the filter is not needed here.

I agree on your topic here regarding the extensibility with filters, but enforcing it is something we should avoid IMHO. That's why I refactored the send function to look something like this:

```
 	final public function send( $hook = null, ...$args ) {


		$request = $this;
		// In case where you don't want to send request.
		if ( $hook ) {
			$request = $this->apply_filters( $hook, ...$args );
		}

		return $this->format_response(
			$this->api_client->send_request( $request )
		);
	}
```

With this approach, if you need a filter, you can pass it and if you don't need it, simple send() will do the trick.

Reply:
> I am aware that filters are optional, but the way you describe their usages is a way where you enforce it.

Yes, enforcing them is the point.

> Adding a filter for a sake of adding it doesn't help too much either.

It might not help too much in most cases, but it could be gold in others, in any case it never hurts. The only downside I could think of is that during tests you’ll need to provide a dummy hook name there.

> What if you need only request without any filters? With the approach, you mentioned here and with the required parameter in send() function make things a bit complicated.

How do you know that you need a request without filters? What does this even mean? Can you give me an example of such a filter, and explain it? With filters, you don’t know who will use them, and when they will use them. You just give them the option.

> you create a filter that will do something and in the cases where you need plain request class, the filter is not needed here.

What is a plain request class? When do you need it?

Reply:
Thank you! In that case let's remove the `if ( $hook ) {` block from the method (it must always be present).

Reply:
Fixed.